### ISSUE TITLE
Use a "fanout" algorithm to speed up search for dirent.

### ISSUE THREAD

**Issue by mgautierfr** [openzim/libzim#424]
- The urlptrpos array list dirents using a sort by namespace and url (`N/url`) order. This allow us to do a binary search on them.
- The parsing of a dirent (at least as we do it now) is "slow". To avoid parsing dirent a lot of time, we use a cache system.
- We have a way (a map now) to allow us to search only on the range associated to the namespace. But with the "removing" of the namespace it is useless as all user content will be put in the same namespace.

The idea here is to use a "fanout" algorithm (I haven't found any documentation on this algorithm but it is what is used (and called) in git format)

The idea is to have a (N sized) array to split the full range into N subranges, and have a fast (O(1)) way to know in which range to search for.

Assuming we have a 256 (2^8) array and the subrange are homogeneous, this allow us to avoid the first 8 steps of the binary search.

I propose to have a 256 array, populated from the first byte of the url.
Each element in the array being the "end" position of the index. (array['f'] is the index of the path "after" pathes starting by `f` (path starting by `g`))
When searching for a the path "foo", we would have to search in the subrange `[array['f'-1], array['f']]`.
By using a simple 256 arrays we reduce the every lookup by 8 steps (if urls are equally spread, which is not).

We could also have more elaborated structure when the array position could point to a range or a subarray (if the range is too big)

```
+--------+---------+-----------+-----------+
|  ....  |   'f'   |    'g'    |    ...    |  <------- First byte of the path
+--------+---------+-----------+-----------+
              |          |
       +------+          +---------+
       v                           v
 +----------------+        +-----+------+-------+------+
 | range [f-g[    |        | ... |  'a' |   'b' |  ... |  <------ Second byte of the path
 +----------------+        +-----+------+-------+------+
                                     |       |      |
                 +-------------------+       |      +----------+
                 v                           v                 v
                ...                         ...               ...   <---- Third byte of the path
```

---

The titleptrpos do the same using `N/title` order with a indirection.
We could do the same thing with the title.
Or not as the main use case for "search by title" is for suggestions and we use xapian when available. (In opposition of path which is the main identifier of the entry and we search *a lot* by path).

**Comment by Jaifroid** [comment]
@mgautierfr I assume this would this require a change in how we read/use dirEntry lookup? Remember Kiwix JS does not use libzim: the reading of indices is fully implemented in JS (not via Emscripten). And also please remember that we *cannot* (yet...) use Xapian, so depend utterly on the title pointer list.

**Comment by mgautierfr** [comment]
No, This algorithm would be implemented at reading. The fanout structure being fill while we are reading dirent. We don't plan for now any change in the way we store the dirent.

In fact, I've even not thought about this. But your comment made me :)
But if we do, it would be some kind of extra structure. The "classic" urlPtrList/TitlePtrList would still be present.

**Comment by mgautierfr** [comment]
Here some more (memory) measurement/estimation about this algorithm.

Measurement are made using the script(s) at https://gist.github.com/mgautierfr/bd786cb987d581efa8af9d5de1fc7ac7

The first script (fanout_algo.py) is the one use to measure the memory usage of this algorithm.
It generates the whole fanout structure and compute the estimated memory usage.
It need a filename listing the fullurl (NS/url) of the zim file (sorted) and the range len after we stop subdividing the ranges.
The output may be big, you probably want to redirect to a file.
`vector_algo.py` does the same but with the algorithm proposed in https://github.com/openzim/libzim/pull/430

The numbers :

I'm using two files:
 - wikipedia_fr_all_mini, 3 657 335 entries, 81Mb to store the urls
 - wikipedia_en_all_maxi, 20_219_212 entries, 556Mb to store the urls

For reference:
- A 256 length range need 8 steps to find a item using binary search
- A 1024 length range need 10 steps to find a item using binary search
- A 3_657_335 length range need 22 steps
- A 20_219_212 length range need 25 steps
- Dividing by 1024 the range length subtract 10 steps

<table>
<tr>
    <th>File | Range len
    <th>Vector
    <th>Fanout
<tr>
	<td>wikipedia_fr_all_mini | 256 (8 steps)
    <td>754Kb
    <td>2205Kb | 2Mb
<tr>
	<td>wikipedia_en_all_maxi | 256
	<td>4676Kb | 4Mb
	<td>11940Kb | 11Mb
<tr>
    <td>wikipedia_fr_all_mini | 1024 (10 steps)
	<td>187Kb
	<td>657Kb
<tr>
    <td>wikipedia_en_all_maxi | 1024
	<td>1168Kb | 1Mb
	<td>3660Kb | 3Mb
<tr>
	<td>wikipedia_fr_all_mini | 3571 (N_entries/1024) (12 steps)
	<td>53Kb
	<td>206Kb
<tr>
	<td>wikipedia_en_all_maxi | 19745 (N_entries/1024) (15 steps)
	<td>61Kb
	<td>251Kb
</table>

We can see that fanout algorithm use 3 to 4 more memory than vector.
But we have to mention that fanout memory usage is the worth case (if we explore all the branches). In real use case, we will build the cache as we need it, the cache will not be totally populated and the cache will not have this size.

On the reading performance (to be measured but without python) :
- the vector algo need a binary search (`log2(N_entries/Rlen)` steps) in the vector to get the final range (which will always be a `Rlen` size).
- the fanout algo need D (depth of the tree) array indexing to get the final range (which will be smaller than `Rlen`)... Assuming the tree is already build. In real case, we will have to build the tree the same time we reading it (but building time will go faster as we advance).

**Comment by Jaifroid** [comment]
Do we have a way of testing "real-world" what the magnitude of the problem is now that namespaces are being removed? Do we have a test ZIM with no namespaces that we could do empirical measurements on?

I think what you propose is ingenious, but please humour me while I try to understand it. Is it something like this?:

* We divide the entire list of URL pointers by a certain number (8 or 16 or 32 or... 256), and we have a list of "key Directory Entries", i.e. the DirEntry that falls at each division point.
* If you know the start and end range within which to search, you have reduced the number of dirEntries to search through by 1/8, or 1/16, 1/32 or 1/256 etc.

It should take very little memory to have a table listing "256 key DirEntries" (like video key frames). That would theoretically be an adequate replacement for namespaces, and could be even faster (as we didn't have 256 namespaces).

Apologies if I've misunderstood the nature of the problem!

**Comment by mgautierfr** [comment]
What you propose is what @veloman-yunkan did in #430 and what I call "vector" algorithm (because the key directory entries are stored in a cpp vector).

As I said in a comment in #430, the "vector" algorithm is "just" a efficient cache system.
We need to to a binary search in the vector to find the range (if we have 1024 key, we need 10 steps) and once we have the range we do the binary search as before (as the range is divided by 1024, we have 10 steps less to do).

On wikipedia fr (3_657_335 entris), with the vector algorithm using 1024 keys, we dividing the full range by 1024 (subranges of 3571 length). It means we do 10 steps of the binary search really quickly, then 12 steps slowly.

Without speedup at all (the basic binary search we have now) we need 22 slow steps.

Whit the fanout algo, assuming the fanout tree is builded and we stop subdividing once we've found a range less than 3517 : We have N array indexing (N depending of the content, the prefix `fullurl[0:N]` must be rare enough to have less than 3517 entries starting with this prefix) and then we do a slow binary search in a range smaller than 3517 (in practice I suppose we may win 1 or 2 steps but not a lot more).

The real performance point is the 10 (if 1024 keys) string comparisons (vector algo) vs the N array indexing (for metadata entries, N=1. For wikipedia en, the worth N value is 69 "Im/osm-intl,12,a,a,270x200.png?lang=en&domain=en.wikipedia.org&title=" while most of the time N is between 5 and 10)

**Comment by mgautierfr** [comment]
I've added https://gist.github.com/mgautierfr/bd786cb987d581efa8af9d5de1fc7ac7#file-fanout_en_all_19745-dump to the gist.
This is the dump of what would be the structure of the fanout tree if we fully create it, using 19745 as max length for a final range.

The beggining is :
```
- b'' (0)
  from 0 to 20219212
  csize : 251Kb | 0Mb
  nb slots : 5
  - b'-' : 0 -> 93 (93)
  - b'A' : 93 -> 14254588 (14254495)
  - b'I' : 14254588 -> 20219199 (5964611)
  - b'M' : 20219199 -> 20219210 (11)
  - b'X' : 20219210 -> 20219212 (2)
  nb subarray : 2
  - b'A' (1)
    from 93 to 14254588
    csize : 163Kb | 0Mb
    nb slots : 107
    - b'!' : 93 -> 313 (220)
    - b'"' : 313 -> 2844 (2531)
    - b'$' : 2844 -> 3140 (296)
    - b'%' : 3140 -> 3165 (25)
    - b'&' : 3165 -> 3232 (67)
    - b"'" : 3232 -> 5468 (2236)
    - b'(' : 5468 -> 9552 (4084)
    [...]
    - b'\xf3' : 14254492 -> 14254588 (96)
    nb subarray : 29
    - b'A1' (2)
      from 16175 to 301044
      csize : 2Kb | 0Mb
      nb slots : 86
      - b'\x00' : 16175 -> 16176 (1)
      - b' ' : 16176 -> 17312 (1136)
      - b'!' : 17312 -> 17313 (1)
      - b'"' : 17313 -> 17318 (5)
      - b'%' : 17318 -> 17350 (32)
      - b'&' : 17350 -> 17355 (5)
      - b"'" : 17355 -> 17365 (10)
      - b'(' : 17365 -> 17371 (6)
      - b'+' : 17371 -> 17423 (52)
      - b',' : 17423 -> 18657 (1234)
      - b'-' : 18657 -> 19640 (983)
      - b'.' : 19640 -> 20192 (552)
      - b'/' : 20192 -> 20454 (262)
      - b'0' : 20454 -> 30926 (10472)
      - b'1' : 30926 -> 35677 (4751)
      - b'2' : 35677 -> 40600 (4923)
```

Which can be read as :

- The first fanout node is for a '' prefix (empty prefix, depth 0).
- It goes from 0 to 20219212 (max entry index)
- It (and its children) would use 251Kb if implemented in C
- It has 5 different slots (the whole furl starts with 5 differents values, namespace in this case).
 - '-' (from 0 to 93) (93 entries in '-' namespace.
 - 'A' (from 93 t o 14254588) (14254495 entries in 'A' namespace).
 - ...
- There is only two slots bigger than 19745, so we need to subdivide only this two slots. (2 subarrays)
- Then we have the node for the 'A' prefix (depth 1)
- It goes from 93 to 14254588 and it would use 163Kb.
- It has 107 slots (urls in 'A' namespace start with 107 different values)
- It has 29 slots to big and we divide them.
- ...
- In the full file, you can see that osm tiles share a long prefix before diverging and so have a long chain of node with only one child until we have enough different values to split the ranges.

It means than :

 - Searching for a entry in `-`, we have only 1 array indexing and then we do a binary search in a 93 length ranges (7 steps)
 - Searching for a url in A namespace starting by `2Y` need 2 array indexing and we binary search in 9 range (4 steps)
 - For `A/Alexh*` (line 1189 in the dump) we would need 6 indexing and we "binary search" in a 1 range
 - For `A/Alexa*` (line 1183) we would need 6 indexing and a binary search in a 17894 range (15 steps)
 - For `I/m/osm-intl,12,a,a,26*` (line 29632) we would need 21 indexing and binary search in a 5 range
 - For `I/m/osm-intl,12,a,a,270x200.png?lang=en&domain=en.wikipedia.org&title=A*` (line 29932) 70 indexing and a binary search in a 1148 range.

**Comment by veloman-yunkan** [comment]
> Here some more (memory) measurement/estimation about this algorithm.

Doing those measurements is helpful but it was somewhat premature. The memory usage of the "vector" approach was not fully optimized. Expect it to be reduced by half or even more

**Comment by mgautierfr** [comment]
This issue is closed by the merge of #430 

But I'm not sure we should consider it as totally closed. The algorithm describe here may still be valid and we may move to it in the future, if needed.
For now we can keep it closed.

**Comment by kelson42** [comment]
@mgautierfr I can only encourage you to push your ideas/comments in a new ticket so this does not get forgotten. We had a successfuly speed-up project, but we might have to come back to this topic at some point in the future.
